libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,324 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/join.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/transform.hpp>
+
+using namespace std;
+using namespace ranges;
+using namespace solidity::frontend;
+
+FunctionCallGraphBuilder::FunctionCallGraphBuilder(ContractDefinition const& _contract):
+	m_contract(&_contract),
+	m_graph(std::make_unique<ContractCallGraph>(_contract))
+{
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::EntryCreation;
+	m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | views::reverse)
+	{
+		for (auto const* stateVar: contract->stateVariables())
+			stateVar->accept(*this);
+
+		for (auto arg: contract->baseContracts())
+			arg->accept(*this);
+
```

Comment:
Maybe we have to reset `m_currentNode` here?

Reply:
Yeah. Probably. I think this is what makes functions called from constructors have an edge from `EntryCreation` rather than from the constructor itself. I left it like that for now because it wasn't outright broken but I was planning to change that once I'm done with the fixes/tests.